app/src/main/java/com/github/moagrius/tileview3/ScrollView.java

```diff
@@ -0,0 +1,550 @@
+   *                       dimension
+   * @return a MeasureSpec integer for the child
+   */
+  public static int getScrollViewChildMeasureSpec(int spec, int padding, int childDimension) {
+    int specMode = MeasureSpec.getMode(spec);
+    int specSize = MeasureSpec.getSize(spec);
+
+    int size = Math.max(0, specSize - padding);
+
+    int resultSize = 0;
+    int resultMode = 0;
+
+    switch (specMode) {
+      case MeasureSpec.EXACTLY:
+        if (childDimension >= 0) {
+          resultSize = childDimension;
+          resultMode = MeasureSpec.EXACTLY;
+        } else if (childDimension == LayoutParams.MATCH_PARENT) {
+          resultSize = size;
+          resultMode = MeasureSpec.EXACTLY;
+        } else if (childDimension == LayoutParams.WRAP_CONTENT) {
+          resultSize = size;
+          resultMode = MeasureSpec.UNSPECIFIED;
+        }
+        break;
+
+      case MeasureSpec.AT_MOST:
+        if (childDimension >= 0) {
+          resultSize = childDimension;
+          resultMode = MeasureSpec.EXACTLY;
+        } else if (childDimension == LayoutParams.MATCH_PARENT) {
+          resultSize = size;
+          resultMode = MeasureSpec.AT_MOST;
+        } else if (childDimension == LayoutParams.WRAP_CONTENT) {
+          resultSize = size;
+          resultMode = MeasureSpec.UNSPECIFIED;
+        }
+        break;
+
+      case MeasureSpec.UNSPECIFIED:
+        if (childDimension >= 0) {
+          resultSize = childDimension;
+          resultMode = MeasureSpec.EXACTLY;
+        } else if (childDimension == LayoutParams.MATCH_PARENT) {
+          resultSize = size;
+          resultMode = MeasureSpec.UNSPECIFIED;
+        } else if (childDimension == LayoutParams.WRAP_CONTENT) {
+          resultSize = size;
+          resultMode = MeasureSpec.UNSPECIFIED;
+        }
+        break;
+    }
+    //noinspection ResourceType
+    return MeasureSpec.makeMeasureSpec(resultSize, resultMode);
+  }
+
+  @Override
+  protected void onLayout(boolean changed, int l, int t, int r, int b) {
+    super.onLayout(changed, l, t, r, b);
+    Log.d("ScrollView", "layout height=" + (b - t) + ", " + getChild().getMeasuredHeight());
+    scrollTo(getScrollX(), getScrollY());
+  }
+
+  /**
+   * Returns whether the ScrollView is currently being flung.
+   *
+   * @return true if the ScrollView is currently flinging, false otherwise.
+   */
+  public boolean isFlinging() {
+    return mIsFlinging;
+  }
+
+  /**
+   * Returns whether the ScrollView is currently being dragged.
+   *
+   * @return true if the ScrollView is currently dragging, false otherwise.
+   */
+  public boolean isDragging() {
+    return mIsDragging;
+  }
+
+  /**
+   * Like {@link View#scrollBy}, but scroll smoothly instead of immediately.
+   *
+   * @param x the number of pixels to scroll by on the X axis
+   * @param y the number of pixels to scroll by on the Y axis
+   */
+  public void smoothScrollBy(int x, int y) {
```

Comment:
that's only going in one direction (y).

i'm +1 on the "don't bother if no child" optimization

also +1 on removing invalidate because it's already called by awakenScrollBars.

+0 (no opinion) on including the maximum scroll (on both axes, not just one) - i'll include it if you think it's valuable